File: apps/push-service/src/main/java/com/example/chat/push/interfaces/rest/SseController.java
package com.example.chat.push.interfaces.rest;

import com.example.chat.push.infrastructure.sender.SseSessionManager;
import lombok.RequiredArgsConstructor;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter;

/**
 * SSE 스트림 컨트롤러
 *
 * 책임: 클라이언트에게 SSE 연결을 제공합니다.
 */
@RestController
@RequestMapping("/api/v1/push")
@RequiredArgsConstructor
public class SseController {
    private static final long SSE_TIMEOUT = 30 * 60 * 1000L; // 30분

    private final SseSessionManager sseSessionManager;

    @GetMapping(value = "/stream/{userId}", produces = MediaType.TEXT_EVENT_STREAM_VALUE)
    public SseEmitter stream(@PathVariable String userId) throws Exception {
        SseEmitter emitter = new SseEmitter(SSE_TIMEOUT);
        emitter.send(SseEmitter.event().name("connect").data("Connected for user: " + userId));
        sseSessionManager.add(userId, emitter);
        return emitter;
    }
}

File: apps/push-service/src/main/kotlin/com/example/chat/push/interfaces/rest/SseController.kt

package com.example.chat.push.interfaces.rest

import com.example.chat.push.infrastructure.sender.SseSessionManager
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter

@RestController
@RequestMapping("/api/v1/push")
class SseController(
    private val sseSessionManager: SseSessionManager
) {
    @GetMapping("/stream/{userId}", produces = [MediaType.TEXT_EVENT_STREAM_VALUE])
    fun stream(
        @PathVariable userId: String
    ): SseEmitter {
        val emitter = SseEmitter(30 * 60 * 1000L) // 30 mins timeout
        // Send initial connect event
        emitter.send(SseEmitter.event().name("connect").data("Connected for user: $userId"))
        sseSessionManager.add(userId, emitter)
        return emitter
    }
}
